### ISSUE TITLE
Ollama Support

### ISSUE THREAD

**Issue by eschaffn** [MaartenGr/BERTopic#1824]
Hi I was wondering if you'd be adding Ollama support, similar to the Llama.cpp support you have?

Thanks!

**Comment by MaartenGr** [comment]
It is not something I am currently considering but it should be relatively straightforward. There are a number of examples [here](https://github.com/MaartenGr/BERTopic/tree/master/bertopic/representation) including the llama.cpp one. 

If you want, a PR would be much appreciated!

**Comment by eschaffn** [comment]
> It is not something I am currently considering but it should be relatively straightforward. There are a number of examples [here](https://github.com/MaartenGr/BERTopic/tree/master/bertopic/representation) including the llama.cpp one.
> 
> If you want, a PR would be much appreciated!

If I get to implementing it myself, I'll submit a PR!

**Comment by elshimone** [comment]
Ollama is now compatible with the OpenAI chat completion API, which means you should be able to use the ChatGPT python client to access Ollama https://ollama.com/blog/openai-compatibility

```
import openai
client = openai.OpenAI(
    base_url = 'http://localhost:11434/v1', #wherever ollama is running
    api_key='ollama', # required, but unused
)

from bertopic.representation import OpenAI
from bertopic import BERTopic

# Create your representation model
representation_model = OpenAI(client, delay_in_seconds=5, model='llama2')

# Use the representation model in BERTopic on top of the default pipeline
topic_model = BERTopic(representation_model=representation_model)
```

Since the openai integration just calls client.chat.completions.create under the hood this should just work.  Alternatively you can use the existing langchain integration to indirectly call through to ollama:

```
from langchain_community.llms import Ollama
from langchain.chains.question_answering import load_qa_chain
from bertopic.representation import LangChain

llm = Ollama(model="llama2")
chain = load_qa_chain(llm, chain_type="stuff")

# Use the representation model in BERTopic on top of the default pipeline
representation_model = LangChain(chain)
topic_model = BERTopic(representation_model=representation_model)
```

**Comment by MaartenGr** [comment]
Thanks for sharing this! I'll make sure to add this to the documentation.

**Comment by elshimone** [comment]
If you point me to the right place in the documentation I can create a PR for this.

**Comment by MaartenGr** [comment]
@elshimone Thanks, that would be great! I imagine the best place for this would be somewhere above/below the `lama.cpp` documentation here: https://maartengr.github.io/BERTopic/getting_started/representation/llm.html#llamacpp

**Comment by eschaffn** [comment]
> Ollama is now compatible with the OpenAI chat completion API, which means you should be able to use the ChatGPT python client to access Ollama https://ollama.com/blog/openai-compatibility
> 
> ```
> import openai
> client = openai.OpenAI(
>     base_url = 'http://localhost:11434/v1', #wherever ollama is running
>     api_key='ollama', # required, but unused
> )
> 
> from bertopic.representation import OpenAI
> from bertopic import BERTopic
> 
> # Create your representation model
> representation_model = OpenAI(client, delay_in_seconds=5, model='llama2')
> 
> # Use the representation model in BERTopic on top of the default pipeline
> topic_model = BERTopic(representation_model=representation_model)
> ```
> 
> Since the openai integration just calls client.chat.completions.create under the hood this should just work. Alternatively you can use the existing langchain integration to indirectly call through to ollama:
> 
> ```
> from langchain_community.llms import Ollama
> from langchain.chains.question_answering import load_qa_chain
> from bertopic.representation import LangChain
> 
> llm = Ollama(model="llama2")
> chain = load_qa_chain(llm, chain_type="stuff")
> 
> # Use the representation model in BERTopic on top of the default pipeline
> representation_model = LangChain(chain)
> topic_model = BERTopic(representation_model=representation_model)
> ```

Hey! Thanks for this, I realized that there's LangChain support for BERTopic which in turn has Ollama integration, so that's what I've been using as well! I should have updated this issue, I apologize.

**Comment by mepearson** [comment]
FYI - Ollama class is deprecated
`The class `Ollama` was deprecated in LangChain 0.3.1 and will be removed in 1.0.0. An updated version of the class exists in the :class:`~langchain-ollama package and should be used instead. To use it run `pip install -U :class:`~langchain-ollama` and import as `from :class:`~langchain_ollama import OllamaLLM``.`

**Comment by zparmley** [comment]
I created an Ollama representation class for my own use, which I hacked up by modifying a copy of the OpenAI representation.  There are a few differences but it's pretty straight forward.  @MaartenGr if you're interested I'd be happy to open a pull request.  Personally I like the simplicity of a dedicated Ollama representation, even if it's not strictly necessary given the other options.

**Comment by MaartenGr** [comment]
@zparmley Yes, that would be great! Would the implementation be of `openai` or using the `ollama` python package? I can imagine that both are valid options here, with the latter providing a bit better support.

**Comment by zparmley** [comment]
@MaartenGr Awesome, I can work on getting the PR together.  Definitely used the `ollama` package when I did this for myself so I'd planned to do the same.
